Add ssh certificate support

42wim opened the pull request:
Fixes #12224 

New KeyType `KeyTypePrincipal` and struct `PrincipalKey` is created.
It's reusing the `PublicKey` table, changes are:
- content/name is the same
- fingerprint is empty
- has `KeyTypePrincipal`
- creates `authorized_principals` file instead of `authorized_keys`

`PublicKey` is modified so it doesn't write KeyTypes with `KeyTypePrincipal` to the authorized_keys.

New section is added to the SSH / GPG keys section.

![image](https://user-images.githubusercontent.com/1810977/87951395-cf551e80-caa8-11ea-98f7-344db6b2361d.png)

The `sshd_config` is also modified by adding a `AuthorizedPrincipalsFile` managed by gitea and a `TrustedUserCAKeys` pointing to `/data/ssh/trusted-user-ca-keys.pem` which should be managed by the gitea administrator. (this contains the publickey of the CA that signs your ssh certificates)

Updates to the original PR:
- built-in ssh server supported
- `SSH_TRUSTED_USER_CA_KEYS` option has been added to the config for specifying your CA keys (and have gitea manage them)
- `SSH_TRUSTED_USER_CA_KEYS_FILENAME` has been added to the config for specifying the filename gitea will manage the CA keys in
- Extra `SSH_AUTHORIZED_PRINCIPALS_ALLOW` has been added to config which by default only allows principals that are equal 
to the user's email or username (to avoid malicious activity). Option to allow anything is also available
- Option to update the authorized principals file has been added to the admin dashboard (like the authorized keys file)

lafriks commented:
Support for this should be also added to built-in ssh server or at least suppor for this should be hidden if built-in ssh server is used

42wim commented:
@lafriks good point, just updated the PR by adding a `ErrSSHBuiltinNotSupported` error and check.

It now show it's not supported when running the builtin server. (and disabled when ssh is disabled)

![image](https://user-images.githubusercontent.com/1810977/87959014-8e620780-cab2-11ea-8388-27d74d888c1f.png)

lafriks commented:
It would be great to add option to manage also CA certs in Gitea admin section to easier add support to built in ssh server later on

42wim commented:
@lafriks yes, but that should be for another PR :)

lafriks commented:
If it is left out it will than later on become breaking change as it would not be possible to migrate automatically to new one

42wim commented:
I've added support for the builtin SSH.

Added a new key `SSH_TRUSTED_USER_CA_KEYS` which expects comma-separated CA keys.
This works for the builtin server now, but the external one still has a separate file.

Should I let gitea write into `/data/ssh/trusted-user-ca-keys.pem` based upon the contents of `SSH_TRUSTED_USER_CA_KEYS` (when the builtin ssh is not used) or should this be the responsibility of the administrator?

lafriks commented:
I would say that for extendability later on it would be better if Gitea would generate that file from setting value

42wim commented:
Updated and the ca file is now getting generated.

lafriks commented on modules/ssh/ssh.go:
```diff
@@ -136,6 +137,52 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
+				},
+			}
+
+			// check the CA of the cert
```
this should be move to L157 for easier code readability

42wim replied:
That's not possible, `c` hasn't been initialized there.

lafriks replied:
Sorry, right :)

silverwind commented on templates/user/settings/keys_principal.tmpl:
```diff
@@ -0,0 +1,65 @@
+				<div class="content">
+					<strong>{{.Name}}</strong>
+					<div class="activity meta">
+						<i>{{$.i18n.Tr "settings.add_on"}} <span>{{.CreatedUnix.FormatShort}}</span> —  <i class="octicon octicon-info"></i> {{if .HasUsed}}{{$.i18n.Tr "settings.last_used"}} <span {{if .HasRecentActivity}}class="green"{{end}}>{{.UpdatedUnix.FormatShort}}</span>{{else}}{{$.i18n.Tr "settings.no_activity"}}{{end}}</i>
```
```suggestion
						<i>{{$.i18n.Tr "settings.add_on"}} <span>{{.CreatedUnix.FormatShort}}</span> —  {{svg "octicon-info" 16}} {{if .HasUsed}}{{$.i18n.Tr "settings.last_used"}} <span {{if .HasRecentActivity}}class="green"{{end}}>{{.UpdatedUnix.FormatShort}}</span>{{else}}{{$.i18n.Tr "settings.no_activity"}}{{end}}</i>
```

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -523,10 +534,11 @@ can_read_info = Read
 hide_openid = Hide from profile
 ssh_disabled = SSH Disabled
-
+ssh_builtin_not_supported = Not supported by builtin SSH
```
This seems unused and if it will be used, I'd reword to `Not supported when using the builtin SSH server`

42wim replied:
good catch, forgot to remove it when I added support, fixed now in new commit

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -489,31 +490,41 @@ keep_email_private_popup = Your email address will be hidden from other users.
 gpg_helper = <strong>Need help?</strong> Have a look at GitHub's guide <a href="%s">about GPG</a>.
 add_new_key = Add SSH Key
 add_new_gpg_key = Add GPG Key
+add_new_principal = Add allowed principal
```
```suggestion
add_new_principal = Add Principal
```

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -489,31 +490,41 @@ keep_email_private_popup = Your email address will be hidden from other users.
+add_new_principal = Add allowed principal
 ssh_key_been_used = This SSH key has already been added to the server.
 ssh_key_name_used = An SSH key with same name is already added to your account.
+ssh_principal_name_used = A SSH certificate principal with the same name is already added to your account.
```
```suggestion
ssh_principal_name_used = A SSH certificate principal with the same name already exists on your account.
```

42wim replied:
No problem with changing that, but it's verbatim from the line above, shouldn't that change too ?

silverwind replied:
Yeah I would suggest changing the line above too, it just sounds better.

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -489,31 +490,41 @@ keep_email_private_popup = Your email address will be hidden from other users.
+ssh_principal_deletion = Remove SSH Certificate Principal
 ssh_key_deletion_desc = Removing an SSH key revokes its access to your account. Continue?
 gpg_key_deletion_desc = Removing a GPG key un-verifies commits signed by it. Continue?
+ssh_principal_deletion_desc = Removing a SSH Certificate Principal revokes it access to your account. Continue?
```
```suggestion
ssh_principal_deletion_desc = Removing a SSH Certificate Principal revokes its access to your account. Continue?
```

silverwind commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,263 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+			return err
+		}
+
+		// .ssh directory should have mode 700, and authorized_keys file should have mode 600.
```
Is that 700 mode to be taken care of here? If not, adjust the comment to not mention it.

42wim replied:
Code for this has been mistakenly removed, added in new commit

silverwind commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,263 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	}
+	defer f.Close()
+
+	// Note: chmod command does not support in Windows.
```
```suggestion
	// Note: chmod is noop on Windows.
```

silverwind commented on custom/conf/app.example.ini:
```diff
@@ -295,9 +298,14 @@ SSH_KEY_TEST_PATH =
+SSH_AUTHORIZED_PRINCIPALS_BACKUP = true
 ; Enable exposure of SSH clone URL to anonymous visitors, default is false
 SSH_EXPOSE_ANONYMOUS = false
+; Specifies the public keys of certificate authorities that are trusted to sign user certificates for authentication.
```
Please elborate which public key format.

42wim replied:
This is literally copied from the sshd_config man page, where it is also not specified. https://manpages.debian.org/experimental/openssh-server/sshd_config.5.en.html#TrustedUserCAKeys

silverwind replied:
Is it `ssh-<format> <key>`? I'd put that as example.

42wim replied:
Yup, it's a normal public key

silverwind commented on custom/conf/app.example.ini:
```diff
@@ -295,9 +298,14 @@ SSH_KEY_TEST_PATH =
 SSH_EXPOSE_ANONYMOUS = false
+; Specifies the public keys of certificate authorities that are trusted to sign user certificates for authentication.
+; Multiple keys should be comma separated.
+SSH_TRUSTED_USER_CA_KEYS =
```
Need to sync all these changes to `docs/content/doc/advanced/config-cheat-sheet.en-us.md` as well.

42wim replied:
Ok, done in latest commit

silverwind commented:
`AuthorizedPrincipalsFile` requires OpenSSH 6.9 or higher which notably excludes CentOS 6 which has 5.3. I don't think we state a minimum OpenSSH version yet, I guess this PR is a good reason to do so.

42wim commented:
We're actually using certificates on centos 6 too :-) 
It's been backported since 2013
```
* Fri Jul 19 2013 Petr Lautrbach <plautrba@redhat.com> 5.3p1-86
  - Add support for certificate key types for users and hosts (#906872)
```
ed25519 support on the other hand..

silverwind commented:
Yeah, I too once manually compiled openssh just for that 😉

zeripath commented on models/ssh_key.go:
```diff
@@ -717,8 +742,11 @@ func RegeneratePublicKeys(t io.StringWriter) error {
 func regeneratePublicKeys(e Engine, t io.StringWriter) error {
 	err := e.Iterate(new(PublicKey), func(idx int, bean interface{}) (err error) {
-		_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
-		return err
+		if bean.(*PublicKey).Type != KeyTypePrincipal {
+			_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
+			return err
+		}
+		return nil
 	})
 	if err != nil {
 		return err
```
```suggestion
	if err := e.Where("type != ?", KeyTypePrincipal).Iterate(new(PublicKey), func(idx int, bean interface{}) (err error) {
		_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
		return err
	}); err != nil {
		return err
```

42wim replied:
Can't seem to commit it with github. "applying suggestions on deleted lines is not supported", but I'll add the change manually

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	}
+
+	for _, key := range keys {
+		if _, err = f.WriteString(key.AuthorizedString()); err != nil {
```
```suggestion
		if key.Type != KeyTypePrincipal {
			continue
		}
		if _, err = f.WriteString(key.AuthorizedString()); err != nil {
```

42wim replied:
That is already implicitly checked, but maybe better to also add this explicit

zeripath replied:
You added the check to the appendPublicKeys function - So for equivalence we should check here too. Either both are implicit or both explicit - having things halfway is a recipe for future bugs.

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	return nil
+}
+
+// GetPrincipalKeyByID returns public key by given ID.
+func GetPrincipalKeyByID(keyID int64) (*PublicKey, error) {
+	return GetPublicKeyByID(keyID)
+}
```
Why does this function exist?

42wim replied:
See line 1197

zeripath replied:
Remove it.

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+		return "", ErrSSHDisabled{}
+	}
+
+	content = strings.TrimRight(content, "\n\r")
+	if strings.ContainsAny(content, "\n\r") {
```
Hmm... `\n\r` feels weird - I appreciate that these functions both just look for any characters in any order but it clangs.

Why not just TrimSpace first instead of TrimRight?

```suggestion
	content = strings.TrimSpace(content)
	if strings.ContainsAny(content, "\r\n") {
```

42wim replied:
This is just copied from the existing functions.
As you can see, the PrincipalKey is almost a copy of PublicKey

42wim replied:
But trimspace is better

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	if strings.ContainsAny(content, "\n\r") {
+		return "", errors.New("only a single line with a single principal please")
+	}
+
+	// remove any unnecessary whitespace now
+	content = strings.TrimSpace(content)
```
If you do accept the earlier trimspace then you can remove these lines

```suggestion
```

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+// deletePrincipalKeys does the actual key deletion but does not update authorized_principals file.
+func deletePrincipalKeys(e Engine, keyIDs ...int64) error {
+	if len(keyIDs) == 0 {
+		return nil
+	}
+
+	_, err := e.In("id", keyIDs).Delete(new(PublicKey))
+	return err
+}
+
+// DeletePrincipalKey deletes SSH key information both in database and authorized_principals file.
+func DeletePrincipalKey(doer *User, id int64) (err error) {
+	key, err := GetPrincipalKeyByID(id)
+	if err != nil {
+		return err
+	}
+
+	// Check if user has access to delete this key.
+	if !doer.IsAdmin && doer.ID != key.OwnerID {
+		return ErrKeyAccessDenied{doer.ID, key.ID, "public"}
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	if err = deletePrincipalKeys(sess, id); err != nil {
+		return err
+	}
+
+	if err = sess.Commit(); err != nil {
+		return err
+	}
+	sess.Close()
+
+	return RewriteAllPrincipalKeys()
+}
+
```
You don't need these functions if you're not going to enforce that the key is a particular type.

Instead at the end of `DeletePublicKey` just check the Type of the key you obtained and do a rewrite:

```go
	if key.Type == KeyTypePrincipal {
		return RewriteAllPrincipalKeys()
	}
```

42wim replied:
Yes, I understand I could do that, but this makes it harder to read where those functions are used, having it differ for principal and publickey makes it more readable imo.

But if you want it removed, I'll do that.

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+}
+
+func regeneratePrincipalKeys(e Engine, t io.StringWriter) error {
+	err := e.Iterate(new(PublicKey), func(idx int, bean interface{}) (err error) {
+		if bean.(*PublicKey).Type == KeyTypePrincipal {
+			_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
```
```suggestion
	if err := e.Where("type = ?", KeyTypePrincipal).Iterate(new(PublicKey), func(idx int, bean interface{}) (err error) {
		_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
		return err
	}); err != nil {
```

zeripath commented on modules/setting/setting.go:
```diff
@@ -711,6 +716,12 @@ func NewContext() {
 		} else if err = os.MkdirAll(SSH.KeyTestPath, 0644); err != nil {
 			log.Fatal("Failed to create '%s': %v", SSH.KeyTestPath, err)
 		}
+
+		trustedUserCaKeys := sec.Key("SSH_TRUSTED_USER_CA_KEYS").Strings(",")
+		if err := ioutil.WriteFile(filepath.Join(SSH.RootPath, "trusted-user-ca-keys.pem"),
+			[]byte(strings.Join(trustedUserCaKeys, "\n")), 0600); err != nil {
+			log.Fatal("Failed to create '%s': %v", filepath.Join(SSH.RootPath, "trusted-user-ca-keys.pem"), err)
+		}
```
Hmm... This might blast away important settings. I'm not sure what to do here. 

Clearly we should only do this if we're going to create the authorized principals file, but we should also only do this if we are sure that we have all of the CA keys that we want to trust.

42wim replied:
I don't understand the problem? What do you mean it might blast away important settings?

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,270 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	if strings.ContainsAny(content, "\r\n") {
+		return "", errors.New("only a single line with a single principal please")
+	}
+
```
We should do some sanity checks to ensure that the certificate is one that we would actually accept - although that could wait for another PR if wanted.

42wim replied:
it's not a certificate, it's just a string, like "myusername"

42wim replied:
I should change the comment on top

zeripath commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,270 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+		_, err = t.WriteString((bean.(*PublicKey)).AuthorizedString())
+		return err
+	}); err != nil {
+	}
```
```suggestion
		return err
	}
```

zeripath commented:
Thanks for this PR! Sorry for the abundance of comments.

Couple of other questions/points:

* Why didn't you add a KeyTypePrincipalDeploy? Or do you think that the certificate route makes these unnecessary?
* We really can't just obliterate the old certificate.pem file - we need to somehow provide an option to do it or some way of appending to it. <- THIS IS A BLOCKER FOR THIS PR
* Given we've decided that the Principals are PublicKeys at the level of the DB then we should make most of the functions work like they are just public keys with a type qualifier. DeletePublicKey should be able to Delete a principal key safely. I don't want to have to remember to check the type and then call a different function -> models/ssh_key should do that work for me.
* We need some way of validating the certificates at addition - similar to the way that we validate the public keys - but if it's not obvious it can be left for another PR I think.

42wim commented:
> * Why didn't you add a KeyTypePrincipalDeploy? Or do you think that the certificate route makes these unnecessary?

Certificates are normally short-lived, imo it wouldn't make sense to use them for deploy keys, which are mostly used in CI/CD or scripts.

> * We really can't just obliterate the old certificate.pem file - we need to somehow provide an option to do it or some way of appending to it. <- THIS IS A BLOCKER FOR THIS PR

This problem must be seen in two different setups, using the docker image and using it without docker.

### Using docker

To be sure of clashes with an existing `~/git/.ssh/trusted-ca-keys.pem` (of which the changes are really slim) we can rename it to `44720ad3-c02b-40e4-9b3a-5e3d37e33cc0-trusted-ca-keys.pem`. This makes sure that we're the one managing it.

### Outside docker

In the real world no-one would actually put the `trusted-ca-keys.pem` in ~/git/.ssh. They would already have a `trusted-ca-keys.pem` in like `/etc/ssh/trusted-ca-keys.pem` and would not let gitea manage that.

So what about adding an extra option `SSH_TRUSTED_USER_CA_KEYS_FILE` which by default is `SSH.RootPath/44720ad3-c02b-40e4-9b3a-5e3d37e33cc0-trusted-user-ca-keys.pem`. 

This way we're sure we are the ones managing it (which works in the docker scenario above) and if they actually want to use the gitea generated file they change their `/etc/ssh/sshd_config` to have `TrustedUserCAKeys` pointed to `SSH.RootPath/44720ad3-c02b-40e4-9b3a-5e3d37e33cc0-trusted-user-ca-keys.pem`. 


> * Given we've decided that the Principals are PublicKeys at the level of the DB then we should make most of the functions work like they are just public keys with a type qualifier. DeletePublicKey should be able to Delete a principal key safely. I don't want to have to remember to check the type and then call a different function -> models/ssh_key should do that work for me.

I do not really agree, to me it's more readable that you're working with "principals" instead of "publickeys", but if this is blocking this PR I'll change it.

> * We need some way of validating the certificates at addition - similar to the way that we validate the public keys - but if it's not obvious it can be left for another PR I think.

I suppose you mean the ones getting put in the `TrustedUserCaKeys`, because the principals itself are just strings that can not be validated. I can add a CA validation probably, will look into that.

zeripath commented:
> > * Why didn't you add a KeyTypePrincipalDeploy? Or do you think that the certificate route makes these unnecessary?
> 
> Certificates are normally short-lived, imo it wouldn't make sense to use them for deploy keys, which are mostly used in CI/CD or scripts.

Cool thanks for clarifying - I've never used SSH Certificates before.

> > * We really can't just obliterate the old certificate.pem file - we need to somehow provide an option to do it or some way of appending to it. <- THIS IS A BLOCKER FOR THIS PR
> 
> ...
> 

Thanks for clarifying - it just looked like a formal name - but if you're happy that we're not clobbering some poor user's standard set-up then we should be able to just use your original name. I was concerned that we would be clobbering and breaking someone's standard set-up in such a way that they could end up not being able to login to their system. 

It might still be best to provide a setting for the name of this file - and not write to it if SSH_TRUSTED_USER_CA_KEYS is empty.  Server owners would be able to deliberately write an empty file by setting `SSH_TRUSTED_USER_CA_KEYS=,`

> > * Given we've decided that the Principals are PublicKeys at the level of the DB then we should make most of the functions work like they are just public keys with a type qualifier. DeletePublicKey should be able to Delete a principal key safely. I don't want to have to remember to check the type and then call a different function -> models/ssh_key should do that work for me.
> 
> I do not really agree, to me it's more readable that you're working with "principals" instead of "publickeys", but if this is blocking this PR I'll change it.

Your current code allows the following:

``go
key, _ := GetPublicKeyByID(x)
_ = DeletePrincipalKey(key)
```

I think that's a very bad idea. Either they enforce that they are working with the correct types, or they should be able to handle all. 

If they just handle all types - (for the appropriate functions) - we can reduce code duplication.

> > * We need some way of validating the certificates at addition - similar to the way that we validate the public keys - but if it's not obvious it can be left for another PR I think.
> 
> ~I suppose you mean the ones getting put in the `TrustedUserCaKeys`, because the principals itself are just strings that can not be validated. I can add a CA validation probably, will look into that.~
> 
> Already did validation of those keys for the internal ssh, now also when writing them to disk for the external SSHD

Sorry I misunderstood how this is supposed to work - I see now that these principals are just usernames/email-addresses. I guess the CA is responsible for checking who gets those principals - however what's to stop a malicious user from adding your principal name to their account before you can get it? That way they could catch your private pushes.

I dunno about this - it might be better if the certificate authority provided certificates with principals for your gitea username@<gitea-host>, or your activated email addresses. You then provide the user with list of options to accept principals for each of those or none-of-those.

42wim commented:
> It might still be best to provide a setting for the name of this file - and not write to it if SSH_TRUSTED_USER_CA_KEYS is empty. Server owners would be able to deliberately write an empty file by setting SSH_TRUSTED_USER_CA_KEYS=,

Agreed

> Sorry I misunderstood how this is supposed to work - I see now that these principals are just usernames/email-addresses. I guess the CA is responsible for checking who gets those principals - however what's to stop a malicious user from adding your principal name to their account before you can get it? That way they could catch your private pushes.

It works the same as the public key, it has the same issues when the malicious user would add your public key to their account before you can get it. 

The SSH certificates basically works as follows: you have a SSH CA somewhere (eg via hashicorp vault), this SSH CA signs your public key and this certificate will contain the principals (that you are allowed to/want to use) and returns you a ssh certificate containing a few things, the specified principals (if you had access to those, normally this is a username) and a specific lifetime.

When you know give this certificate to ssh, it'll check:
- if you own the private key used for this certificate
- if the lifetime is valid
- if it was signed by the correct SSH CA
- and if the principal exists somewhere in the `authorized_principals` file.

The SSH CA part / signing is something done outside of gitea, just like the creating of the public/private keys is.

42wim commented:
I think I handled all your comments @zeripath

42wim commented:
Is there something else I need to do before this can be merged?

lunny commented on custom/conf/app.example.ini:
```diff
@@ -295,9 +298,21 @@ SSH_KEY_TEST_PATH =
 SSH_KEYGEN_PATH = ssh-keygen
 ; Enable SSH Authorized Key Backup when rewriting all keys, default is true
-SSH_BACKUP_AUTHORIZED_KEYS = true
+SSH_AUTHORIZED_KEYS_BACKUP = true
```
Since the change, it's a break change.

42wim replied:
I was correcting the documentation as `SSH_BACKUP_AUTHORIZED_KEYS` isn't correct. (it has nothing to do with my change, I can remove it again if you like)

See https://github.com/go-gitea/gitea/blob/master/modules/setting/setting.go#L120 on master

zeripath replied:
@lunny he's correct here: https://github.com/go-gitea/gitea/blob/95d478c29c60c38c17dcaadeda0192979ea0d451/modules/setting/setting.go#L722

It's been SSH_AUTHORIZED_KEYS_BACKUP for 3 years!

zeripath commented on models/ssh_key.go:
```diff
@@ -36,9 +36,10 @@ import (
+	tplCommentPrefix         = `# gitea public key`
+	tplCommand               = "%s --config=%q serv key-%d"
+	tplPublicKey             = tplCommentPrefix + "\n" + `command=%q,no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty %s` + "\n"
+	authorizedPrincipalsFile = "authorized_principals"
```
Add a blank line here

```suggestion

		authorizedPrincipalsFile = "authorized_principals"
```

techknowlogick commented:
Thanks for this PR, I'm trying to test this however I am new to using SSH certificates. Could you provide a brief getting started on using SSH certificates in Gitea?

zeripath commented on options/locale/locale_en-US.ini:
```diff
@@ -492,33 +493,43 @@ keep_email_private_popup = Your email address will be hidden from other users.
 manage_gpg_keys = Manage GPG Keys
 add_key = Add Key
 ssh_desc = These public SSH keys are associated with your account. The corresponding private keys allow full access to your repositories.
+principal_desc = These allowed SSH certificate principals are associated with your account. The corresponding principals allow full access to your repositories.
```
```suggestion
principal_desc = These SSH certificate principals are associated with your account and allow full access to your repositories.
```

zeripath commented on options/locale/locale_en-US.ini:
```diff
@@ -492,33 +493,43 @@ keep_email_private_popup = Your email address will be hidden from other users.
 ssh_key_been_used = This SSH key has already been added to the server.
-ssh_key_name_used = An SSH key with same name is already added to your account.
+ssh_key_name_used = An SSH key with same name already exists on your account.
+ssh_principal_name_used = A SSH certificate principal with the same name already exists on your account.
```
This appears redundant.

zeripath commented on options/locale/locale_en-US.ini:
```diff
@@ -492,33 +493,43 @@ keep_email_private_popup = Your email address will be hidden from other users.
+principal_content = Content
 add_key_success = The SSH key '%s' has been added.
 add_gpg_key_success = The GPG key '%s' has been added.
+add_principal_success = The allowed SSH certificate principal '%s' has been added.
```
```suggestion
add_principal_success = The SSH certificate principal '%s' has been added.
```

zeripath commented:
I'm still very concerned that allowing users to add principals - essentially usernames - they will accept without some way of asserting that they control these principals is a security hazard.

In particular your current code:

```go
	// Key name of same user cannot be duplicated.
	has, err := sess.
		Where("content = ? AND name = ? AND type = ?", content, name, KeyTypePrincipal).
		Get(new(PublicKey))
	if err != nil {
		return nil, err
	} else if has {
		return nil, ErrKeyAlreadyExist{0, "", content}
	}
```

Would allow a situation whereby a malicious Eve could racily add principals that Alice owns certificates to before Alice has a chance to add them. Alice would then add her principals - they would be accepted - but when Alice then pushes her changes they go to Eve's repositories. Couple this with PushToCreate and Eve could steal private repositories. 

If you change to use my suggestion then, Eve would simply be denying Alice her principal - which is a little better but far from perfect. However a Trudy could use the information to figure out which principals are in use - I guess that's OK.

The ordering of principal presentation could also be affected - say Alice has two certificates A & B but only has B added as a principal to Gitea. Chuck then adds A as a prinicipal and suddenly Alice is Chuck by accident. 

I don't know how to fix this - you would really need some way of activating the certificate principal to really make this safe.

Either that or you fall back to only accepting principals that match the username or activated email addresses.

42wim commented:
@zeripath but the same issue exists with ssh public keys right? you can find public keys of users on github (or other places) add them to your account and it's the same attack vector?

> Would allow a situation whereby a malicious Eve could racily add principals that Alice owns certificates to before Alice has a chance to add them. Alice would then add her principals - they would be accepted - but when Alice then pushes her changes they go to Eve's repositories. Couple this with PushToCreate and Eve could steal private repositories.

I'm not sure how it could racily add the principals? If Eve added the one of Alice before Alice and Alice will try to add them, they would not be accepted, Alice will get a pop-up show it's already in use.

42wim commented:
@techknowlogick  quick tutorial, normally you would use something as vault in production for managing this stuff.

This example is using the gitea built-in ssh server

## generate ssh-ca used for signing

```
$ ssh-keygen -t ed25519 -N '' -C ssh-ca -f ssh-ca
Generating public/private ed25519 key pair.
Your identification has been saved in ssh-ca.
Your public key has been saved in ssh-ca.pub.
The key fingerprint is:
SHA256:bOBEoJxWWRAQVdwNxWpjfHxRbfCBLbCEhDJKLA3YtR8 ssh-ca
The key's randomart image is:
+--[ED25519 256]--+
|.o*=BO+.+*oo..=+ |
|.o.B.=....+ oo o+|
|  B o E. o . ....|
| . . + +* o .    |
|      ooSo .     |
|       .         |
|                 |
|                 |
|                 |
+----[SHA256]-----+
```

## put the ca public key in gitea config
```
$ cat ssh-ca.pub
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIDAwkb1+7mK5BnwWXxl8Flh3mG716I0eXfhDBc9TOEzn ssh-ca
```

So in your gitea `app.ini` put

`SSH_TRUSTED_USER_CA_KEYS = ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIDAwkb1+7mK5BnwWXxl8Flh3mG716I0eXfhDBc9TOEzn ssh-ca`

restart gitea

## generate your own sshkey that will be used for authentication

(No password for quick testing, obviously do not do this in production)

```
$ ssh-keygen -t ed25519 -N ''  -f mykey
Generating public/private ed25519 key pair.
Your identification has been saved in mykey.
Your public key has been saved in mykey.pub.
The key fingerprint is:
SHA256:5ac8VPYwb6MYu5WharbN/hIjWRe7Xu51XhXw1Xu2sYM wim@server
The key's randomart image is:
+--[ED25519 256]--+
|             .  o|
|            . o o|
|          . =o o.|
|         o.oo= o+|
|        So+.o.*.*|
|        oooB.E.=.|
|         .B++o  =|
|        o+.+. .oo|
|       ooo=o... .|
+----[SHA256]-----+
```

## sign this sshkey with the ssh-ca 

with principal `wim` and a lifetime of 1 hour

```
$ ssh-keygen -s ssh-ca -I keyid -n wim -V -5:+1h -z 1 mykey.pub
Signed user key mykey-cert.pub: id "keyid" serial 1 for wim valid from 2020-08-29T13:59:05 to 2020-08-29T14:59:10
```

## start an ssh-agent

$ eval `ssh-agent`
Agent pid 42183

## add the key and certificate to your agent

```
$ ssh-add mykey
Identity added: mykey (wim@server)
Certificate added: mykey-cert.pub (keyid)
```

`ssh-add -l` should show them both

```
$ ssh-add -l
256 SHA256:5ac8VPYwb6MYu5WharbN/hIjWRe7Xu51XhXw1Xu2sYM wim@server (ED25519)
256 SHA256:5ac8VPYwb6MYu5WharbN/hIjWRe7Xu51XhXw1Xu2sYM wim@server (ED25519-CERT)
```

## in gitea add the principal `wim` to your account

![image](https://user-images.githubusercontent.com/1810977/91636650-8aef6380-ea02-11ea-90c2-a7a51983ea99.png)

## now clone from your gitea instance

Just use git clone and you should be able by using your certificate.

```
git clone git@yourserver:youruser/yourproject.git
```

42wim commented on models/ssh_key.go:
```diff
@@ -1030,3 +1058,225 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+		return nil, err
+	}
+
+	// Key name of same user cannot be duplicated.
+	has, err := sess.
+		Where("content = ? AND name = ? AND type = ?", content, name, KeyTypePrincipal).
+		Get(new(PublicKey))
+	if err != nil {
+		return nil, err
+	} else if has {
+		return nil, ErrKeyAlreadyExist{0, "", content}
+	}
```
You're right, I handled that in the caller which puts the content in name and content, so it's identical.

line 55 of keys.go
```
	if _, err = models.AddPrincipalKey(ctx.User.ID, content, content, 0); err != nil {
```
 
But I can change it here too if you want

42wim commented:
@zeripath I've added a commit that should fix your security comments (as discussed on discord too)

This commit adds a SSH_AUTHORIZED_PRINCIPALS_ALLOW which is default
`email,username` this means that users only can add the principals
that match their email or username.

To allow anything the admin need to set the option `anything`.

This allows for a safe default in gitea which protects against malicious
users using other user's prinicipals. (before that user could set it).

This commit also has some small other fixes from the last code review.

lafriks commented:
Lint is failing

42wim commented:
> Lint is failing

Was fixed in my editor but not saved on disk apparently, should be fixed now

42wim commented:
I've got also the code for an extra feature `SSH_AUTHORIZED_PRINCIPALS_ALLOW_AUTO`.

This would allow you to specify `username` or `email` (like `SSH_AUTHORIZED_PRINCIPALS_ALLOW`) and it'll update the database and add the specified principal (user's email/username) to each user and every new created user. (the downside is that if the admin chooses to remove this, the principals are still available in the database)

(AFAIK there's no easy way then adding it in the database instead of checking it at runtime because of the way the `serv` thing works. To do it at runtime would mean a considerable amount of changes to the code)

Should I add this feature to this PR also or keep it for a follow-up one?

zeripath commented:
@42wim I've added an AuthorizedPrincipalsCommand to this. I hope you don't mind and wonder if you'd be able to just check it. I've also allowed any activated email address to be set.

There's one last thing that I think we need to look at - the errors coming out of the principals code do not appear to be localisable. Now, there's lots of bits of the above SSH normal key code that aren't localisable but I don't know what people think about adding more.

I'm ambivalent - if you think it would be too much work to fix and can tell me that the AuthorizedPrincipalCommand thing works I'll approve.

42wim commented on modules/setting/setting.go:
```diff
@@ -736,7 +736,7 @@ func NewContext() {
 	SSH.AuthorizedPrincipalsAllow = sec.Key("SSH_AUTHORIZED_PRINCIPALS_ALLOW").Strings(",")
 	if len(SSH.AuthorizedPrincipalsAllow) == 0 {
-		SSH.AuthorizedPrincipalsAllow = []string{"email", "username"}
+		SSH.AuthorizedPrincipalsAllow = []string{"username", "email"}
```
what's the rationale for this change?

zeripath replied:
usernames are quicker to check than emails

42wim commented:
> I've added an AuthorizedPrincipalsCommand to this. I hope you don't mind and wonder if you'd be able to just check it. I've also allowed any activated email address to be set.

@zeripath 
This won't work

```
AuthorizedPrincipalsCommand /path/to/gitea keys-principal -e git -u %u -i %i
```
the `%i` is the ssh certificate keyid, it's not the principals, this key id isn't in gitea (it's something the ssh-ca generates).


The right way I think is to do this:
```
AuthorizedPrincipalsCommand /path/to/gitea keys-principal -e git -u %u
```

And have it spit out all the `KeyTypePrincipals` for that specific user, because all those are valid.

silverwind commented on models/ssh_key.go:
```diff
@@ -1041,3 +1092,202 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+}
+
+func rewriteAllPrincipalKeys(e Engine) error {
+	//Don't rewrite key if internal server
```
```suggestion
	// Don't rewrite key if internal server
```

42wim commented:
As discussed on discord, it's not possible to do this with the `AuthorizedPrincipalsCommand` I think, but if the reason for this is performance, it looks like it's not really an issue with a big authorized_principals file.

I just did a test with a file containing 1k, 10k,100k, 1mio and 10mio entries

These are first hit results and when the actual principal isn't present, so every line has to be searched.
On 2nd hit, it keeps a steady 0m0260s

```
10 entries: real    0m0.254s
1k entries: real    0m0.260s
10k entries: real    0m0.282s
100k entries: real    0m0.312s
1mio entries: real    0m0.498s
10mio entries: real    0m13.315s (2GB file)
```
running this: 
`time git clone ssh://wim@localhost:3001/wim/testing.git`

zeripath commented:
OK i've removed the authorizedprincipalcommand thing as it would require a large amount of work to make work.

Still need a decision on the localisation error issue.

42wim commented:
About localisation, maybe for another PR that fixes it for everything ssh related ?

zeripath commented:
We're almost there @42wim... I think we need to have this functionality off unless it is explicitly switched on. Therefore I propose the below patch:

<details>
<summary>Patch</summary>

```patch
From d89c05a69faaf61f0fe2ce6ba9e8e56007f7e723 Mon Sep 17 00:00:00 2001
From: Andrew Thornton <art27@cantab.net>
Date: Sun, 6 Sep 2020 14:56:20 +0100
Subject: [PATCH] By default do not show principal options unless there are CA
 keys set or they are explicitly set

Signed-off-by: Andrew Thornton <art27@cantab.net>
---
 custom/conf/app.example.ini                   |  12 +-
 .../doc/advanced/config-cheat-sheet.en-us.md  |   8 +-
 modules/setting/setting.go                    |  79 ++++++++----
 routers/user/setting/keys.go                  |   4 +
 templates/user/settings/keys_principal.tmpl   | 114 +++++++++---------
 5 files changed, 126 insertions(+), 91 deletions(-)

diff --git a/custom/conf/app.example.ini b/custom/conf/app.example.ini
index de5819130..c93c62440 100644
--- a/custom/conf/app.example.ini
+++ b/custom/conf/app.example.ini
@@ -306,15 +306,15 @@ SSH_KEY_TEST_PATH =
 SSH_KEYGEN_PATH = ssh-keygen
 ; Enable SSH Authorized Key Backup when rewriting all keys, default is true
 SSH_AUTHORIZED_KEYS_BACKUP = true
-; Enable SSH Authorized Principals Backup when rewriting all keys, default is true
-SSH_AUTHORIZED_PRINCIPALS_BACKUP = true
-; Enable exposure of SSH clone URL to anonymous visitors, default is false
-SSH_AUTHORIZED_PRINCIPALS_ALLOW = email, username
 ; Determines which principals to allow
+; - empty: if SSH_TRUSTED_USER_CA_KEYS is empty this will default to off, otherwise will default to email, username.
+; - off: Do not allow authorized principals
 ; - email: the principal must match the user's email
 ; - username: the principal must match the user's username
 ; - anything: there will be no checking on the content of the principal
-SSH_EXPOSE_ANONYMOUS = false
+SSH_AUTHORIZED_PRINCIPALS_ALLOW = email, username
+; Enable SSH Authorized Principals Backup when rewriting all keys, default is true
+SSH_AUTHORIZED_PRINCIPALS_BACKUP = true
 ; Specifies the public keys of certificate authorities that are trusted to sign user certificates for authentication.
 ; Multiple keys should be comma separated.
 ; E.g."ssh-<algorithm> <key>". or "ssh-<algorithm> <key1>, ssh-<algorithm> <key2>".
@@ -325,6 +325,8 @@ SSH_TRUSTED_USER_CA_KEYS =
 ; If you're running your own ssh server and you want to use the gitea managed file you'll also need to modify your
 ; sshd_config to point to this file. The official docker image will automatically work without further configuration.
 SSH_TRUSTED_USER_CA_KEYS_FILENAME =
+; Enable exposure of SSH clone URL to anonymous visitors, default is false
+SSH_EXPOSE_ANONYMOUS = false
 ; Indicate whether to check minimum key size with corresponding type
 MINIMUM_KEY_SIZE_CHECK = false
 ; Disable CDN even in "prod" mode
diff --git a/docs/content/doc/advanced/config-cheat-sheet.en-us.md b/docs/content/doc/advanced/config-cheat-sheet.en-us.md
index 7376fbb03..c327d58e3 100644
--- a/docs/content/doc/advanced/config-cheat-sheet.en-us.md
+++ b/docs/content/doc/advanced/config-cheat-sheet.en-us.md
@@ -198,11 +198,11 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `SSH_PORT`: **22**: SSH port displayed in clone URL.
 - `SSH_LISTEN_HOST`: **0.0.0.0**: Listen address for the built-in SSH server.
 - `SSH_LISTEN_PORT`: **%(SSH\_PORT)s**: Port for the built-in SSH server.
-- `SSH_TRUSTED_USER_CA_KEYS`: **\<empty\>**: Specifies the public keys of certificate authorities that are trusted to sign user certificates for authentication. Multiple keys should be comma separated. E.g.`ssh-<algorithm> <key>` or `ssh-<algorithm> <key1>, ssh-<algorithm> <key2>`. For more information see `TrustedUserCAKeys` in the sshd config man pages. When empty no file will be created.
+- `SSH_TRUSTED_USER_CA_KEYS`: **\<empty\>**: Specifies the public keys of certificate authorities that are trusted to sign user certificates for authentication. Multiple keys should be comma separated. E.g.`ssh-<algorithm> <key>` or `ssh-<algorithm> <key1>, ssh-<algorithm> <key2>`. For more information see `TrustedUserCAKeys` in the sshd config man pages. When empty no file will be created and `SSH_AUTHORIZED_PRINCIPALS_ALLOW` will default to `off`.
 - `SSH_TRUSTED_USER_CA_KEYS_FILENAME`: **`RUN_USER`/.ssh/gitea-trusted-user-ca-keys.pem**: Absolute path of the `TrustedUserCaKeys` file gitea will manage. If you're running your own ssh server and you want to use the gitea managed file you'll also need to modify your sshd_config to point to this file. The official docker image will automatically work without further configuration.
-- `SSH_CREATE_AUTHORIZED_PRINCIPALS_FILE`: **true**: Gitea will create a authorized_principals file by default when it is not using the internal ssh server. If you intend to use the AuthorizedKeysCommand functionality then you should turn this off.
-- `SSH_AUTHORIZED_PRINCIPALS_BACKUP`: **true**: Enable SSH Authorized Principals Backup when rewriting all keys, default is true.
-- `SSH_AUTHORIZED_PRINCIPALS_ALLOW`: **username, email**: \[username, email, anything\]: Specify the principals values that users are allowed to use as principal. When set to `anything` no checks are done on the principal string.
+- `SSH_AUTHORIZED_PRINCIPALS_ALLOW`: **off** or **username, email**: \[off, username, email, anything\]: Specify the principals values that users are allowed to use as principal. When set to `anything` no checks are done on the principal string. When set to `off` authorized principal are not allowed to be set.
+- `SSH_CREATE_AUTHORIZED_PRINCIPALS_FILE`: **false/true**: Gitea will create a authorized_principals file by default when it is not using the internal ssh server and `SSH_AUTHORIZED_PRINCIPALS_ALLOW` is not `off`.
+- `SSH_AUTHORIZED_PRINCIPALS_BACKUP`: **false/true**: Enable SSH Authorized Principals Backup when rewriting all keys, default is true if `SSH_AUTHORIZED_PRINCIPALS_ALLOW` is not `off`.
 - `OFFLINE_MODE`: **false**: Disables use of CDN for static files and Gravatar for profile pictures.
 - `DISABLE_ROUTER_LOG`: **false**: Mute printing of the router log.
 - `CERT_FILE`: **https/cert.pem**: Cert file path used for HTTPS. From 1.11 paths are relative to `CUSTOM_PATH`.
diff --git a/modules/setting/setting.go b/modules/setting/setting.go
index 40f26f6a0..93554182a 100644
--- a/modules/setting/setting.go
+++ b/modules/setting/setting.go
@@ -708,6 +708,49 @@ func NewContext() {
 		SSH.StartBuiltinServer = false
 	}
 
+	trustedUserCaKeys := sec.Key("SSH_TRUSTED_USER_CA_KEYS").Strings(",")
+	for _, caKey := range trustedUserCaKeys {
+		pubKey, _, _, _, err := gossh.ParseAuthorizedKey([]byte(caKey))
+		if err != nil {
+			log.Fatal("Failed to parse TrustedUserCaKeys: %s %v", caKey, err)
+		}
+
+		SSH.TrustedUserCAKeysParsed = append(SSH.TrustedUserCAKeysParsed, pubKey)
+	}
+	if len(trustedUserCaKeys) > 0 {
+		// Set the default as email,username otherwise we can leave it empty
+		sec.Key("SSH_AUTHORIZED_PRINCIPALS_ALLOW").MustString("email,username")
+	} else {
+		sec.Key("SSH_AUTHORIZED_PRINCIPALS_ALLOW").MustString("off")
+	}
+
+	SSH.AuthorizedPrincipalsAllow = sec.Key("SSH_AUTHORIZED_PRINCIPALS_ALLOW").Strings(",")
+	pos := 0
+	anything := false
+authorizedPrinicipalLoop:
+	for _, value := range SSH.AuthorizedPrincipalsAllow {
+		v := strings.ToLower(strings.TrimSpace(value))
+		switch v {
+		case "off":
+			pos = 0
+			break authorizedPrinicipalLoop
+		case "email":
+			fallthrough
+		case "username":
+			SSH.AuthorizedPrincipalsAllow[pos] = v
+			pos++
+		case "anything":
+			anything = true
+		}
+	}
+	if pos == 0 {
+		SSH.AuthorizedPrincipalsAllow = []string{"off"}
+	} else if anything {
+		SSH.AuthorizedPrincipalsAllow = []string{"anything"}
+	} else {
+		SSH.AuthorizedPrincipalsAllow = SSH.AuthorizedPrincipalsAllow[:pos]
+	}
+
 	if !SSH.Disabled && !SSH.StartBuiltinServer {
 		if err := os.MkdirAll(SSH.RootPath, 0700); err != nil {
 			log.Fatal("Failed to create '%s': %v", SSH.RootPath, err)
@@ -715,16 +758,7 @@ func NewContext() {
 			log.Fatal("Failed to create '%s': %v", SSH.KeyTestPath, err)
 		}
 
-		trustedUserCaKeys := sec.Key("SSH_TRUSTED_USER_CA_KEYS").Strings(",")
-
-		for _, caKey := range trustedUserCaKeys {
-			_, _, _, _, err := gossh.ParseAuthorizedKey([]byte(caKey))
-			if err != nil {
-				log.Fatal("Failed to parse TrustedUserCaKeys: %s %v", caKey, err)
-			}
-		}
-
-		if len(trustedUserCaKeys) > 0 {
+		if len(trustedUserCaKeys) > 0 && SSH.AuthorizedPrincipalsAllow[0] != "off" {
 			fname := sec.Key("SSH_TRUSTED_USER_CA_KEYS_FILENAME").MustString(filepath.Join(SSH.RootPath, "gitea-trusted-user-ca-keys.pem"))
 			if err := ioutil.WriteFile(fname,
 				[]byte(strings.Join(trustedUserCaKeys, "\n")), 0600); err != nil {
@@ -733,11 +767,6 @@ func NewContext() {
 		}
 	}
 
-	SSH.AuthorizedPrincipalsAllow = sec.Key("SSH_AUTHORIZED_PRINCIPALS_ALLOW").Strings(",")
-	if len(SSH.AuthorizedPrincipalsAllow) == 0 {
-		SSH.AuthorizedPrincipalsAllow = []string{"email", "username"}
-	}
-
 	SSH.MinimumKeySizeCheck = sec.Key("MINIMUM_KEY_SIZE_CHECK").MustBool()
 	minimumKeySizes := Cfg.Section("ssh.minimum_key_sizes").Keys()
 	for _, key := range minimumKeySizes {
@@ -747,22 +776,20 @@ func NewContext() {
 			delete(SSH.MinimumKeySizes, strings.ToLower(key.Name()))
 		}
 	}
+
 	SSH.AuthorizedKeysBackup = sec.Key("SSH_AUTHORIZED_KEYS_BACKUP").MustBool(true)
-	SSH.AuthorizedPrincipalsBackup = sec.Key("SSH_AUTHORIZED_PRINCIPALS_BACKUP").MustBool(true)
 	SSH.CreateAuthorizedKeysFile = sec.Key("SSH_CREATE_AUTHORIZED_KEYS_FILE").MustBool(true)
-	SSH.CreateAuthorizedPrincipalsFile = sec.Key("SSH_CREATE_AUTHORIZED_PRINCIPALS_FILE").MustBool(true)
-	SSH.ExposeAnonymous = sec.Key("SSH_EXPOSE_ANONYMOUS").MustBool(false)
 
-	trustedUserCaKeys := sec.Key("SSH_TRUSTED_USER_CA_KEYS").Strings(",")
-	for _, caKey := range trustedUserCaKeys {
-		pubKey, _, _, _, err := gossh.ParseAuthorizedKey([]byte(caKey))
-		if err != nil {
-			log.Fatal("Failed to parse TrustedUserCaKeys: %s %v", caKey, err)
-		}
-
-		SSH.TrustedUserCAKeysParsed = append(SSH.TrustedUserCAKeysParsed, pubKey)
+	if SSH.AuthorizedPrincipalsAllow[0] == "off" {
+		SSH.AuthorizedPrincipalsBackup = false
+		SSH.CreateAuthorizedPrincipalsFile = false
+	} else {
+		SSH.AuthorizedPrincipalsBackup = sec.Key("SSH_AUTHORIZED_PRINCIPALS_BACKUP").MustBool(true)
+		SSH.CreateAuthorizedPrincipalsFile = sec.Key("SSH_CREATE_AUTHORIZED_PRINCIPALS_FILE").MustBool(true)
 	}
 
+	SSH.ExposeAnonymous = sec.Key("SSH_EXPOSE_ANONYMOUS").MustBool(false)
+
 	sec = Cfg.Section("server")
 	if err = sec.MapTo(&LFS); err != nil {
 		log.Fatal("Failed to map LFS settings: %v", err)
diff --git a/routers/user/setting/keys.go b/routers/user/setting/keys.go
index 0c73b9512..baaa60cdd 100644
--- a/routers/user/setting/keys.go
+++ b/routers/user/setting/keys.go
@@ -23,6 +23,7 @@ func Keys(ctx *context.Context) {
 	ctx.Data["PageIsSettingsKeys"] = true
 	ctx.Data["DisableSSH"] = setting.SSH.Disabled
 	ctx.Data["BuiltinSSH"] = setting.SSH.StartBuiltinServer
+	ctx.Data["AllowPrincipals"] = (len(setting.SSH.AuthorizedPrincipalsAllow) > 0 && setting.SSH.AuthorizedPrincipalsAllow[0] != "off")
 
 	loadKeysData(ctx)
 
@@ -33,6 +34,9 @@ func Keys(ctx *context.Context) {
 func KeysPost(ctx *context.Context, form auth.AddKeyForm) {
 	ctx.Data["Title"] = ctx.Tr("settings")
 	ctx.Data["PageIsSettingsKeys"] = true
+	ctx.Data["DisableSSH"] = setting.SSH.Disabled
+	ctx.Data["BuiltinSSH"] = setting.SSH.StartBuiltinServer
+	ctx.Data["AllowPrincipals"] = (len(setting.SSH.AuthorizedPrincipalsAllow) > 0 && setting.SSH.AuthorizedPrincipalsAllow[0] != "off")
 
 	if ctx.HasError() {
 		loadKeysData(ctx)
diff --git a/templates/user/settings/keys_principal.tmpl b/templates/user/settings/keys_principal.tmpl
index a8c05fd4e..c163263ea 100644
--- a/templates/user/settings/keys_principal.tmpl
+++ b/templates/user/settings/keys_principal.tmpl
@@ -1,65 +1,67 @@
-<h4 class="ui top attached header">
-	{{.i18n.Tr "settings.manage_ssh_principals"}}
-	<div class="ui right">
-	{{if not .DisableSSH}}
-		<div class="ui blue tiny show-panel button" data-panel="#add-ssh-principal-panel">{{.i18n.Tr "settings.add_new_principal"}}</div>
-	{{else}}
-		<div class="ui blue tiny button disabled">{{.i18n.Tr "settings.ssh_disabled"}}</div>
-	{{end}}
-	</div>
-</h4>
-<div class="ui attached segment">
-	<div class="ui key list">
-		<div class="item">
-			{{.i18n.Tr "settings.principal_desc"}}
+{{if .AllowPrincipals}}
+	<h4 class="ui top attached header">
+		{{.i18n.Tr "settings.manage_ssh_principals"}}
+		<div class="ui right">
+		{{if not .DisableSSH}}
+			<div class="ui blue tiny show-panel button" data-panel="#add-ssh-principal-panel">{{.i18n.Tr "settings.add_new_principal"}}</div>
+		{{else}}
+			<div class="ui blue tiny button disabled">{{.i18n.Tr "settings.ssh_disabled"}}</div>
+		{{end}}
 		</div>
-		{{range .Principals}}
+	</h4>
+	<div class="ui attached segment">
+		<div class="ui key list">
 			<div class="item">
-				<div class="right floated content">
-					<button class="ui red tiny button delete-button" id="delete-principal" data-url="{{$.Link}}/delete?type=principal" data-id="{{.ID}}">
-						{{$.i18n.Tr "settings.delete_key"}}
-					</button>
-				</div>
-				<i class="big send icon {{if .HasRecentActivity}}green{{end}}" {{if .HasRecentActivity}}data-content="{{$.i18n.Tr "settings.principal_state_desc"}}" data-variation="inverted tiny"{{end}}></i>
-				<div class="content">
-					<strong>{{.Name}}</strong>
-					<div class="activity meta">
-						<i>{{$.i18n.Tr "settings.add_on"}} <span>{{.CreatedUnix.FormatShort}}</span> —  {{svg "octicon-info" 16}} {{if .HasUsed}}{{$.i18n.Tr "settings.last_used"}} <span {{if .HasRecentActivity}}class="green"{{end}}>{{.UpdatedUnix.FormatShort}}</span>{{else}}{{$.i18n.Tr "settings.no_activity"}}{{end}}</i>
+				{{.i18n.Tr "settings.principal_desc"}}
+			</div>
+			{{range .Principals}}
+				<div class="item">
+					<div class="right floated content">
+						<button class="ui red tiny button delete-button" id="delete-principal" data-url="{{$.Link}}/delete?type=principal" data-id="{{.ID}}">
+							{{$.i18n.Tr "settings.delete_key"}}
+						</button>
+					</div>
+					<i class="big send icon {{if .HasRecentActivity}}green{{end}}" {{if .HasRecentActivity}}data-content="{{$.i18n.Tr "settings.principal_state_desc"}}" data-variation="inverted tiny"{{end}}></i>
+					<div class="content">
+						<strong>{{.Name}}</strong>
+						<div class="activity meta">
+							<i>{{$.i18n.Tr "settings.add_on"}} <span>{{.CreatedUnix.FormatShort}}</span> —  {{svg "octicon-info" 16}} {{if .HasUsed}}{{$.i18n.Tr "settings.last_used"}} <span {{if .HasRecentActivity}}class="green"{{end}}>{{.UpdatedUnix.FormatShort}}</span>{{else}}{{$.i18n.Tr "settings.no_activity"}}{{end}}</i>
+						</div>
 					</div>
 				</div>
-			</div>
-		{{end}}
+			{{end}}
+		</div>
 	</div>
-</div>
-<br>
+	<br>
 
-<div {{if not .HasPrincipalError}}class="hide"{{end}} id="add-ssh-principal-panel">
-	<h4 class="ui top attached header">
-		{{.i18n.Tr "settings.add_new_principal"}}
-	</h4>
-	<div class="ui attached segment">
-		<form class="ui form" action="{{.Link}}" method="post">
-			{{.CsrfTokenHtml}}
-			<div class="field {{if .Err_Content}}error{{end}}">
-				<label for="content">{{.i18n.Tr "settings.principal_content"}}</label>
-				<input id="ssh-principal-content" name="content" value="{{.content}}" autofocus required>
-			</div>
-			<input name="title" type="hidden" value="principal">
-			<input name="type" type="hidden" value="principal">
-			<button class="ui green button">
-				{{.i18n.Tr "settings.add_new_principal"}}
-			</button>
-		</form>
+	<div {{if not .HasPrincipalError}}class="hide"{{end}} id="add-ssh-principal-panel">
+		<h4 class="ui top attached header">
+			{{.i18n.Tr "settings.add_new_principal"}}
+		</h4>
+		<div class="ui attached segment">
+			<form class="ui form" action="{{.Link}}" method="post">
+				{{.CsrfTokenHtml}}
+				<div class="field {{if .Err_Content}}error{{end}}">
+					<label for="content">{{.i18n.Tr "settings.principal_content"}}</label>
+					<input id="ssh-principal-content" name="content" value="{{.content}}" autofocus required>
+				</div>
+				<input name="title" type="hidden" value="principal">
+				<input name="type" type="hidden" value="principal">
+				<button class="ui green button">
+					{{.i18n.Tr "settings.add_new_principal"}}
+				</button>
+			</form>
+		</div>
 	</div>
-</div>
 
-<div class="ui small basic delete modal" id="delete-principal">
-	<div class="ui icon header">
-		<i class="trash icon"></i>
-		{{.i18n.Tr "settings.ssh_principal_deletion"}}
-	</div>
-	<div class="content">
-		<p>{{.i18n.Tr "settings.ssh_principal_deletion_desc"}}</p>
+	<div class="ui small basic delete modal" id="delete-principal">
+		<div class="ui icon header">
+			<i class="trash icon"></i>
+			{{.i18n.Tr "settings.ssh_principal_deletion"}}
+		</div>
+		<div class="content">
+			<p>{{.i18n.Tr "settings.ssh_principal_deletion_desc"}}</p>
+		</div>
+		{{template "base/delete_modal_actions" .}}
 	</div>
-	{{template "base/delete_modal_actions" .}}
-</div>
+{{end}}
-- 
2.25.1
```

</details>


[0001-By-default-do-not-show-principal-options-unless-ther.txt](https://github.com/go-gitea/gitea/files/5179741/0001-By-default-do-not-show-principal-options-unless-ther.txt)

42wim commented:
@zeripath ok for me, makes sense.

Only checking for the "off" is enough, not? code could be a bit more readable like this

```go
if valueInSlice("off", SSH.AuthorizedPrincipalsAllow) {
   SSH.AuthorizedPrincipalsAllow="off"
}
```

```go
func valueInSlice(value string, slice []string) bool {
	for _, v := range slice {
		v = strings.ToLower(strings.TrimSpace(v))
		if v == value {
			return true
		}
	}

	return false
}
```


Also the `len(setting.SSH.AuthorizedPrincipalsAllow) > 0` check can be removed because it always will be > 0
(that check also isn't done in the `setting.go`)

Or we could also add another bool only used internally  `SSH.AuthorizedPrincipalsEnabled`

techknowlogick commented:
I've been able to run through this (thanks for instructions @42wim) using the built-in SSH server and it works as expected, and I don't have feedback for the code, still need to run through with opensshd testing first before +1

@42wim (and @zeripath) thanks for all your hard work on this PR :)

6543 commented:
@techknowlogick did you had time to run with opensshd?

techknowlogick commented on modules/setting/setting.go:
```diff
@@ -718,8 +752,17 @@ func NewContext() {
 	SSH.AuthorizedKeysBackup = sec.Key("SSH_AUTHORIZED_KEYS_BACKUP").MustBool(true)
 	SSH.CreateAuthorizedKeysFile = sec.Key("SSH_CREATE_AUTHORIZED_KEYS_FILE").MustBool(true)
+
+	SSH.AuthorizedPrincipalsBackup = false
```
I just updated this if statement, as previously it was `if AuthorizedPrincipalsEnabled then hardcode false` whereas now it will accept settings if principals are enabled.

techknowlogick commented:
I'm 99% sure I'm doing something wrong when testing this using opensshd in docker, but am having troubles figuring out what I'm doing wrong. Is anyone else able to test this scenario?

42wim commented:
@techknowlogick it was the 1% my fault, sorry :( 
I made a typo in the docker sshd_config template. It should work correctly now.

lunny commented on models/ssh_key.go:
```diff
@@ -1041,3 +1070,202 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	if err = sess.Commit(); err != nil {
+		return nil, err
+	}
+
```
Add `sess.Close()` here is better.

lafriks replied:
there is already `defer sess.Close()`
https://github.com/go-gitea/gitea/pull/12281/files#diff-bc1f3a75a0a9d450f654cec20fd672e9R1084

zeripath replied:
I suspect the potential problem is that rewriteallpublickeys et al will open another session so if you don't close you could end up keeping things around?

lunny replied:
@lafriks I know that. This is not conflicted. An extra `Close` will keep only one connection at the same moment if `rewriteallpublickeys` will create some session.

lafriks commented on models/ssh_key.go:
```diff
@@ -1041,3 +1070,202 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	if err = sess.Commit(); err != nil {
+		return nil, err
+	}
+
```
ok, got it

```suggestion

	sess.Close()
```

6543 commented:
pleace resolve conflicts :)

lafriks commented:
@techknowlogick can you review again please?

6543 commented:
It would be nice to ad some Docs about ssh-certificate (tld howto and some "read more" links ...)

42wim commented:
@6543 something like https://github.com/go-gitea/gitea/pull/12281#issuecomment-683283290 ?
Where should I add those docs?

6543 commented:
@42wim would be a nice to have :+1:

6543 reviewed (approved):
nice ...!

still some docs like you described in your comment would be better ...

techknowlogick commented:
docs could be a different PR

6543 commented:
@42wim :point_up: :)
